packages/rspack-plugin-react-refresh/client/reactRefreshEntry.js

```diff
@@ -1,4 +1,50 @@
-const RefreshRuntime = require("react-refresh/runtime");
+/**
+ * The following code is modified based on
+ * https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/0b960573797bf38926937994c481e4fec9ed8aa6/client/ReactRefreshEntry.js
+ *
+ * MIT Licensed
+ * Author Michael Mok
+ * Copyright (c) 2019 Michael Mok
+ * https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/0b960573797bf38926937994c481e4fec9ed8aa6/LICENSE
+ */
 
-RefreshRuntime.injectIntoGlobalHook(self);
-self.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
+var RefreshRuntime = require("react-refresh/runtime");
+var safeThis = (function () {
+	// copied from core-js-pure/features/global-this
+	"use strict";
+
+	var check = function (it) {
+		return it && it.Math == Math && it;
+	};
+
+	// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
+	// eslint-disable-next-line es/no-global-this -- safe
+	return (
+		check(typeof globalThis == "object" && globalThis) ||
+		check(typeof window == "object" && window) ||
+		// eslint-disable-next-line no-restricted-globals -- safe
+		check(typeof self == "object" && self) ||
+		check(typeof __webpack_require__.g == "object" && __webpack_require__.g) ||
+		// eslint-disable-next-line no-new-func -- fallback
+		(function () {
+			return this;
+		})() ||
+		this ||
+		Function("return this")()
+	);
+})();
+
+if (process.env.NODE_ENV !== "production") {
```

Comment:
does webpack needs this toos?

Reply:
Yes, it's just for defensive programming to avoid it being included in the production bundle.

Reply:
However, react fast refresh plugin will have a runtime error in production mode. So I cannot give you a test for that.

Reply:
https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/0b960573797bf38926937994c481e4fec9ed8aa6/client/ReactRefreshEntry.js#L6 they really have this 🤦